src/v/security/gssapi.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "bytes/bytes.h"
+
+#include <gssapi/gssapi.h>
+#include <gssapi/gssapi_ext.h>
+
+#include <string_view>
+
+namespace security::gss {
+
+class buffer_view {
+public:
+    explicit buffer_view(::gss_buffer_desc bd)
+      : _impl{bd} {}
+    explicit buffer_view(std::string_view sv)
+      : _impl{sv.size(), const_cast<char*>(sv.data())} {}
+    explicit buffer_view(bytes_view bv)
+      : _impl{bv.size(), const_cast<unsigned char*>(bv.data())} {}
+
+    explicit operator std::string_view() const {
+        return {static_cast<char*>(_impl.value), _impl.length};
+    };
+    explicit operator bytes_view() const {
+        return {static_cast<unsigned char*>(_impl.value), _impl.length};
+    };
+
+    const void* value() const { return _impl.value; }
+    size_t size() const { return _impl.length; }
+    const gss_buffer_desc* operator&() const { return &_impl; };
+    // TODO BP: Should I wrap every call that only uses this as an input param?
+    gss_buffer_desc* operator&() { return &_impl; };
+
+private:
+    ::gss_buffer_desc _impl;
+};
```

Comment:
Does it need to be explicit that `buffer_view` doesn't 'own' the `gss_buffer_desc` that it contains (unlike `buffer` below)?  What if there's a call to `gss_release_buffer` on a copy of the buffer that `buffer_view` contains?

Reply:
The `buffer_view` is non-owning, like `std::string_view`, typically used as an adaptor for input parameters to GSSAPI calls. The type makes it explicit, did you have something else in mind?

The conversion operator can't be const because c. Wrapping all of the calls and `const_cast`ing is a fair bit of effort, they're mostly called just once.

Reply:
Oh ok the suffix `_view` implies non-ownership then?  I haven't worked much with `std::string_view` so maybe that's the piece I was missing.  Appreciate the explanation.

Reply:
I guess. We also have `bytes_view` https://github.com/redpanda-data/redpanda/blob/dev/src/v/bytes/bytes.h#L33, which is why there are conversion operators to `std::string_view` and `bytes_view`. The standard has `std::span` for similar, but we don't really use that much, so I just went with `_view` suffix.